### PR TITLE
Ledger ETH support for BIP44 account paths

### PR DESCRIPTION
**Pull request by kaladinlight**

### CODE REVIEW

File: packages/hdwallet-portis/src/portis.ts [hunk 1]
```diff
@@ -181,7 +181,7 @@ export class PortisHDWallet implements HDWallet, ETHWallet, BTCWallet {
     return this.info.describePath(msg)
   }
 
-  public async getPublicKeys(msg: GetPublicKey[]): Promise<PublicKey[]> {
```
Comment (elmutt) [inline packages/hdwallet-portis/src/portis.ts]: Why change portis?

Comment (kaladinlight) [reply packages/hdwallet-portis/src/portis.ts]: The function definition in the HDWallet interface changed.

Comment (elmutt) [reply packages/hdwallet-portis/src/portis.ts]: less risk to just leave it

Comment (kaladinlight) [reply packages/hdwallet-portis/src/portis.ts]: Should behave exactly the same way.

Comment (elmutt) [reply packages/hdwallet-portis/src/portis.ts]: It changed adding `| null`

no need to touch portis please

Comment (kaladinlight) [reply packages/hdwallet-portis/src/portis.ts]: Right, it matches HDWallet definition in core.

Comment (elmutt) [reply packages/hdwallet-portis/src/portis.ts]: its less risk to not touch this line

Comment (keepkeyjon) [reply packages/hdwallet-portis/src/portis.ts]: @elmutt it'll be fine, no risk. Types in typescript are just an annotation that the compiler checks, so if it builds, it's good.

File: packages/hdwallet-ledger/src/ethereum.ts [hunk 2]
```diff
@@ -1,45 +1,84 @@
...
   const res = await transport.call('Eth', 'getAddress', bip32path, !!msg.showDisplay)
   handleError(res, transport, 'Unable to obtain ETH address from device.')
 
   return res.payload.address
 }
 
-export async function ethSignTx(transport: LedgerTransport, msg: ETHSignTx): Promise<ETHSignedTx> {
-  const bip32path = addressNListToBIP32(msg.addressNList)
+// Adapted from https://github.com/LedgerHQ/ledger-wallet-webtool
+export async function ethGetPublicKeys (transport: LedgerTransport, msg: Array<core.GetPublicKey>): Promise<Array<core.PublicKey>> {
```
Comment (keepkeyjon) [inline packages/hdwallet-ledger/src/ethereum.ts]:
```suggestion
export async function ethGetPublicKeys (transport: LedgerTransport, msg: Array<core.GetPublicKey>): Promise<Array<core.PublicKey | null>> {
```

File: packages/hdwallet-ledger/src/bitcoin.ts [hunk 3]
```diff
@@ -79,6 +64,56 @@ export async function btcGetAddress (transport: LedgerTransport, msg: BTCGetAddr
   return res.payload.bitcoinAddress
 }
 
+// Adapted from https://github.com/LedgerHQ/ledger-wallet-webtool
+export async function btcGetPublicKeys (transport: LedgerTransport, msg: Array<core.GetPublicKey>): Promise<Array<core.PublicKey>> {
```
Comment (keepkeyjon) [inline packages/hdwallet-ledger/src/bitcoin.ts]:
```suggestion
export async function btcGetPublicKeys (transport: LedgerTransport, msg: Array<core.GetPublicKey>): Promise<Array<core.PublicKey | null>> {
```

File: packages/hdwallet-ledger/src/ethereum.ts [hunk 4]
```diff
@@ -84,22 +123,26 @@ export function ethSupportsNativeShapeShift(): boolean {
   return false
 }
 
-export function ethGetAccountPaths (msg: ETHGetAccountPath): Array<ETHAccountPath> {
+export function ethGetAccountPaths (msg: core.ETHGetAccountPath): Array<core.ETHAccountPath> {
+  /* TODO support both ETH account paths
```
Comment (keepkeyjon) [inline packages/hdwallet-ledger/src/ethereum.ts]: You might have to comment out a corresponding test case.

File: packages/hdwallet-ledger/src/bitcoin.ts [hunk 5]
```diff
@@ -79,6 +64,56 @@ export async function btcGetAddress (transport: LedgerTransport, msg: BTCGetAddr
...
+      publicKey = compressPublicKey(publicKey)
+
+      const coinDetails: any = networksUtil[core.slip44ByCoin(coin)]
+      const childNum: number = addressNList[addressNList.length -1]
+
+      scriptType = scriptType || core.BTCInputScriptType.SpendAddress
+      const networkMagic = coinDetails.bitcoinjs.bip32.public[scriptType]
+
+      const xpub = createXpub(
+        3,
```
Comment (keepkeyjon) [inline packages/hdwallet-ledger/src/bitcoin.ts]:
```suggestion
        addressNList.length,
```

File: packages/hdwallet-ledger/src/ethereum.ts [hunk 6]
```diff
@@ -1,45 +1,84 @@
...
+
+      let { payload: { publicKey, chainCode } } = res2
+      publicKey = compressPublicKey(publicKey)
+
+      const coinDetails: any = networksUtil[core.slip44ByCoin(coin)]
+      const childNum: number = addressNList[addressNList.length -1]
+      const networkMagic = coinDetails.bitcoinjs.bip32.public[scriptType]
+
+      const xpub = createXpub(
+        3,
```
Comment (keepkeyjon) [inline packages/hdwallet-ledger/src/ethereum.ts]:
```suggestion
        addressNList.length,
```

Comment (keepkeyjon) [reply packages/hdwallet-ledger/src/ethereum.ts]: (this was broken before your change, but that's what it really should be)

File: packages/hdwallet-core/src/wallet.ts [hunk 7]
```diff
@@ -19,7 +19,8 @@ export interface GetPublicKey {
   showDisplay?: boolean,
   scriptType?: BTCInputScriptType,
   curve: string,
-  coin?: Coin
+  coin?: Coin,
+  symbol?: Symbol
```
Comment (keepkeyjon) [inline packages/hdwallet-core/src/wallet.ts]: You should key this off of what the `coin` is, instead of adding another one.

Comment (kaladinlight) [reply packages/hdwallet-core/src/wallet.ts]: Symbol is already on the payload we are passing in which is why I added that as an optional key

Comment (keepkeyjon) [reply packages/hdwallet-core/src/wallet.ts]: yeah, I see why you did that now... lemme think about it for a bit.

File: packages/hdwallet-ledger/src/ethereum.ts (hunk 4 above)
Comment (kaladinlight) [reply packages/hdwallet-ledger/src/ethereum.ts]: I added one to just test the single path and added skip to the other one until the non bip44 is supported.

Comment (keepkeyjon) [reply packages/hdwallet-ledger/src/ethereum.ts]: 👍

File: packages/hdwallet-ledger/src/ethereum.ts [hunk 8]
```diff
@@ -1,45 +1,84 @@
...
-export async function ethSignTx(transport: LedgerTransport, msg: ETHSignTx): Promise<ETHSignedTx> {
-  const bip32path = addressNListToBIP32(msg.addressNList)
+// Adapted from https://github.com/LedgerHQ/ledger-wallet-webtool
+export async function ethGetPublicKeys (transport: LedgerTransport, msg: Array<core.GetPublicKey>): Promise<Array<core.PublicKey>> {
+    const xpubs = []
+
+    for (const getPublicKey of msg) {
+      let { addressNList, coin, symbol, scriptType } = getPublicKey
+
+      if (symbol !== 'ETH') {
```
Comment (keepkeyjon) [inline packages/hdwallet-ledger/src/ethereum.ts]: I think this test should match what their device actually cares about, which I think is that `addressNList` must start with  `[0x80000000 + 44, 0x80000000 + 60]` if you're using the Eth app. Then we can avoid adding `symbol` to `GetPublicKey`, which seems odd (not sure why it's being passed in that other spot, but it shouldn't be what we're using here).

File: packages/hdwallet-ledger/src/ledger.ts [hunk 9]
```diff
@@ -389,54 +380,20 @@ export class LedgerHDWallet implements core.HDWallet, core.BTCWallet, core.ETHWa
...
 
-      const xpub = createXpub(
-        3,
-        fingerprint,
-        childNum,
-        chainCode,
-        publicKey,
-        networkMagic
-      )
+    const { name } = res.payload
```
Comment (dviramontes) [inline packages/hdwallet-ledger/src/ledger.ts]:
```suggestion
    const { payload: { name } } = res
```

File: packages/hdwallet-ledger/src/ethereum.ts [hunk 10]
```diff
@@ -72,7 +111,7 @@ export async function ethSignTx(transport: LedgerTransport, msg: ETHSignTx): Pro
     v: parseInt(v, 16),
     r: '0x' + r,
     s: '0x' + s,
-    serialized: '0x' + toHexString(tx.serialize())
+    serialized: '0x' + core.toHexString(tx.serialize())
```
Comment (dviramontes) [inline packages/hdwallet-ledger/src/ethereum.ts]:
```suggestion
    serialized: `0x${core.toHexString(tx.serialize())}`
```

File: packages/hdwallet-ledger/src/ethereum.ts (hunk 4 above)
Comment (kaladinlight) [reply packages/hdwallet-ledger/src/ethereum.ts]: Decided to leave the tests and `ethGetAccountPaths` alone (including both ledger live and chrome derivation paths). I will just add the length to the `ethGetPublicKeys` filter to only include BIP44 account path for now.

Comment (keepkeyjon) [reply packages/hdwallet-ledger/src/ethereum.ts]: even better

File: packages/hdwallet-ledger/src/ethereum.ts (hunk 10 above)
Comment (kaladinlight) [reply packages/hdwallet-ledger/src/ethereum.ts]: This syntax is spread throughout the whole code base. Not going to worry about this one spot at this time.

File: packages/hdwallet-ledger/src/ethereum.ts [hunk 11]
```diff
@@ -1,45 +1,85 @@
...
-  const bip32path = addressNListToBIP32(msg.addressNList)
+// Adapted from https://github.com/LedgerHQ/ledger-wallet-webtool
+export async function ethGetPublicKeys (transport: LedgerTransport, msg: Array<core.GetPublicKey>): Promise<Array<core.PublicKey | null>> {
+    const xpubs = []
+
+    for (const getPublicKey of msg) {
+      let { addressNList, coin, scriptType } = getPublicKey
+
+      // Only get public keys for ETH account paths
+      if (!addressNList.includes(0x80000000 + 44, 0) || !addressNList.includes(0x80000000 + 60, 1)) {
```
Comment (keepkeyjon) [inline packages/hdwallet-ledger/src/ethereum.ts]: woah, didn't know about that second arg to `includes()`, nice!

**Comment by keepkeyjon** [comment]
@kaladinlight mind adding release notes in changelog.md?

**Pull request merged**